Add score checksum calculations

Lekuruu opened the pull request:
After some digging around, I've finally found the "formula" for score checksums. So far, it's been showing promising results, but it still needs to be tested extensively.

Lekuruu committed:
Add "BooleanToInteger" helper function

Lekuruu committed:
Implement replay checksum calculation

ascenttree commented on common/replays.go:
```diff
@@ -132,6 +136,30 @@ func (header *ReplayHeader) Grade() Grade {
+			strconv.Itoa(int(header.MaxCombo)),
+			strconv.Itoa(int(BooleanToInteger(header.FullCombo))),
+			strconv.Itoa(int(math.Round(header.TotalScore))),
+			strconv.Itoa(int(header.Grade())),
```
```suggestion
			header.Grade(),
```

Lekuruu replied:
header.Grade is an integer

ascenttree replied:
..what

ascenttree replied:
nevermind then

ascenttree commented on common/serialization.go:
```diff
@@ -84,3 +84,10 @@ func ReadU16(data []byte) uint16 {
 	return uint8(data[0])
 }
+
+func BooleanToInteger(value bool) uint8 {
```
Can we also refactor the stream module to use this function now?

Lekuruu replied:
Sure, go ahead

Lekuruu committed:
Add a partial implementation of the leaderboard handlers

Co-authored-by: tree <tree@ascension.wtf>

Lekuruu committed:
Add replay download endpoint

Co-authored-by: tree <tree@ascension.wtf>

Lekuruu committed:
Add score date & username to leaderboard response

Co-authored-by: tree <tree@ascension.wtf>

Lekuruu committed:
Fix unused import

Lekuruu committed:
Make project usable again

Lekuruu committed:
Upgrade to go 1.25.8

Lekuruu committed:
Attempt to implement score checksum validation

Lekuruu committed:
Fix dockerfile

ascenttree committed:
Finish login packets

Lekuruu committed:
Validate mods data length

Lekuruu committed:
Add `QByteArray` reader & writer / Add replay-specific datetime reader &
writer

Lekuruu committed:
Implement replay checksums properly / refactor replay serialization &
include proper mods parsing

This seems to work fine with replays, I kinda just have to test it for
score submission too, since I'm not sure if score sub uses the same
checksum algorithm.
The naming is also something to improve, e.g. for
"CreateModsChecksumToken"

Lekuruu committed:
`ReplayDateTime` -> `QDateTime`

Lekuruu committed:
Fix score checksums in score submission

Lekuruu committed:
Merge branch 'main' into score-checksums
